src/runtime/security/negotiation_manager.cpp

```diff
@@ -38,80 +40,100 @@ inline bool in_white_list(task_code code)
     return is_negotiation_message(code) || fd::is_failure_detector_message(code);
 }
 
-negotiation_map negotiation_service::_negotiations;
-utils::rw_lock_nr negotiation_service::_lock;
+negotiation_map negotiation_manager::_negotiations;
+utils::rw_lock_nr negotiation_manager::_lock;
 
-negotiation_service::negotiation_service() : serverlet("negotiation_service") {}
+negotiation_manager::negotiation_manager() : serverlet("negotiation_manager") {}
 
-void negotiation_service::open_service()
+void negotiation_manager::open_service()
 {
     register_rpc_handler_with_rpc_holder(
-        RPC_NEGOTIATION, "Negotiation", &negotiation_service::on_negotiation_request);
+        RPC_NEGOTIATION, "Negotiation", &negotiation_manager::on_negotiation_request);
 }
 
-void negotiation_service::on_negotiation_request(negotiation_rpc rpc)
+void negotiation_manager::on_negotiation_request(negotiation_rpc rpc)
 {
     dassert(!rpc.dsn_request()->io_session->is_client(),
-            "only server session receive negotiation request");
+            "only server session receives negotiation request");
 
     // reply SASL_AUTH_DISABLE if auth is not enable
     if (!security::FLAGS_enable_auth) {
         rpc.response().status = negotiation_status::type::SASL_AUTH_DISABLE;
         return;
     }
 
-    server_negotiation *srv_negotiation = nullptr;
-    {
-        utils::auto_read_lock l(_lock);
-        srv_negotiation =
-            static_cast<server_negotiation *>(_negotiations[rpc.dsn_request()->io_session].get());
+    std::shared_ptr<negotiation> nego = get_negotiation(rpc);
+    if (nullptr != nego) {
+        server_negotiation *srv_negotiation = static_cast<server_negotiation *>(nego.get());
+        srv_negotiation->handle_request(rpc);
     }
+}
 
-    dassert(srv_negotiation != nullptr,
-            "negotiation is null for msg: {}",
-            rpc.dsn_request()->rpc_code().to_string());
-    srv_negotiation->handle_request(rpc);
+void negotiation_manager::on_negotiation_response(error_code err, negotiation_rpc rpc)
+{
+    dassert(rpc.dsn_request()->io_session->is_client(),
+            "only client session receives negotiation response");
+
+    std::shared_ptr<negotiation> nego = get_negotiation(rpc);
```

Comment:
Why use shared_ptr here... The memory of `nego` is still owned by `_negotiations`. You are transferring the memory to std::shared_ptr... And when this function completes, `nego` will free the internal memory because no one reference it.

I don't understand why such simple mistake happens but tests still pass.

Reply:
Add test something like the follows:

```
TEST(negotiation_manager_test, client_negotiation) {
  negotiation_rpc fake_rpc;
  ... // mock data
  on_negotiation_response(ERR_OK, fake);

  // ensure not null, ensure memory valid.
  ASSERT_EQ(get_negotiation(rpc), xxx);
}
```

Reply:
I think we must use shared_ptr here. because if we use unique_ptr, the negotiation maybe destroyed by `negotiation_manager::on_rpc_disconnected`.  so in L77 we must declare a shared_ptr to add it's ref count